man/vpoll.3
.\" Copyright (C) 2019 VirtualSquare. Project Leader: Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" as published by the Free Software Foundation, either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.
.\"
.\" Automatically generated by Pandoc 3.1.11
.\"
.TH "VPOLL" "3" "January 2024" "VirtualSquare" "Library Functions Manual"
.SH NAME
vpoll_create, vpoll_ctl, vpoll_close \- generate synthetic events for
poll/select/ppoll/pselect/epoll
.SH SYNOPSIS
\f[CB]#include *vpoll.h*\f[R]
.PP
\f[CB]int vpoll_create(uint32_t\f[R]
\f[I]init_events\f[R]\f[CB], int\f[R] \f[I]flags\f[R]\f[CB]);\f[R]
.PP
\f[CB]int vpoll_ctl(int\f[R] \f[I]fd\f[R]\f[CB], int\f[R]
\f[I]op\f[R]\f[CB], uint32_t\f[R] \f[I]events\f[R]\f[CB]);\f[R]
.PP
\f[CB]int vpoll_close(int\f[R] \f[I]fd\f[R]\f[CB]);\f[R]
.SH DESCRIPTION
This library permits one to create a vpoll file descriptor
\[lq]vpollfd\[rq] that can be used in poll/select/ppoll/pselect/epoll(2)
system calls.
The events reported by a can be controlled by \f[CB]vpoll_ctl\f[R].
\f[CB]vpoll\f[R] encodes the events using the same flags EPOLL* defined
in \f[CB]epoll_ctl(2)\f[R].
.PP
The library uses the vpoll extension for eventfd if the kernel provides
it or the vpoll device implemented by the vpoll kernel module.
When neither of the kernel supports are available the libvpoll library
(partially) emulates the vpoll feature using socketpair(2).
This emulation supports only \f[CB]EPOLLIN\f[R], \f[CB]EPOLLOUT\f[R]
flags and a non standard version of EPOLLHUP/EPOLLRDHUP.
.TP
\f[CB]vpoll_create\f[R]
This function creates a \[lq]vpollfd\[rq].
The argument \f[I]init_events\f[R] is used to set the initial state of
events.
The following value can be included in \f[I]flags\f[R]:
.TP
\f[CB]\f[R]
\f[CB]FD_CLOEXEC\f[R]:
Set the close\-on\-exec flag on the new file descriptor.
See the description of the O_CLOEXEC flag in open(2) for reasons why
this may be useful.
.TP
\f[CB]vpoll_ctl\f[R]
This function changes the set of pending events reported by a
\[lq]vpollfd\[rq].
The argument \f[I]op\f[R] can take the following values:
.TP
\f[CB]\f[R]
\f[CB]VPOLL_CTL_ADDEVENTS\f[R]:
the events set in the argument \f[I]events\f[R] are added to the set of
pending events.
.TP
\f[CB]\f[R]
\f[CB]VPOLL_CTL_DELEVENTS\f[R]:
the events set in the argument \f[I]events\f[R] are deleted from the set
of pending events.
.TP
\f[CB]\f[R]
\f[CB]VPOLL_CTL_SETEVENTS\f[R]:
the value of the argument \f[I]events\f[R] is assigned to the set of
pending events.
.TP
\f[CB]vpoll_close\f[R]
This function closes the vpoll file descritor.
.SH RETURN VALUE
\f[CB]vpoll_create\f[R] returns the new file descriptor, or \-1 if an
error occurred (in which case, errno is set appropriately)
.PP
\f[CB]vpoll_ctl\f[R] and \f[CB]vpoll_close\f[R] return zero in case of
success.
On error, \-1 is returned, and errno is set appropriately.
.SH EXAMPLE
.IP
.EX
#define _GNU_SOURCE
#include *stdio.h*
#include *stdlib.h*
#include *unistd.h*
#include *fcntl.h*
#include *errno.h*
#include *sys/epoll.h*
#include *vpoll.h*

int main(int argc, char *argv[]) {
  int vpollfd = vpoll_create(0, FD_CLOEXEC);
  int epfd = epoll_create1(EPOLL_CLOEXEC);
  struct epoll_event reqevents={EPOLLIN | EPOLLRDHUP | EPOLLERR |
      EPOLLOUT | EPOLLHUP | EPOLLPRI};
  epoll_ctl(epfd,  EPOLL_CTL_ADD, vpollfd, &reqevents);
  switch (fork()) {
      case 0:
          sleep(3);
          vpoll_ctl(vpollfd, VPOLL_CTL_ADDEVENTS,  EPOLLIN);
          sleep(3);
          vpoll_ctl(vpollfd, VPOLL_CTL_ADDEVENTS,  EPOLLIN);
          sleep(3);
          vpoll_ctl(vpollfd, VPOLL_CTL_ADDEVENTS,  EPOLLOUT);
          sleep(3);
          vpoll_ctl(vpollfd, VPOLL_CTL_ADDEVENTS,  EPOLLHUP);
          sleep(3);
          exit(0);
      default:
          while (1) {
              struct epoll_event ev;
              int n = epoll_wait(epfd, &ev, 1, 1000);
              if (n * 0) {
                  perror(\[dq]epoll_wait\[dq]);
                  break;
              }
              if (n * 0) {
                  printf(\[dq]GOT event %x\[rs]n\[dq], ev.events);
                  vpoll_ctl(vpollfd, VPOLL_CTL_DELEVENTS, ev.events);
                  if (ev.events & EPOLLHUP)
                          break;
              } else {
                  printf(\[dq]timeout\[rs]n\[dq]);
              }
          }
          break;
      case \-1:
          printf(\[dq]fork error\[rs]n\[dq]);
  }
  vpoll_close(vpollfd);
  close(epfd);
  return 0;
}
.EE
.PP
On a machine running a Linux Kernel providing eventfd/vpoll or the vpoll
device the output of this program is:
.IP
.EX
timeout
timeout
GOT event 1
timeout
timeout
GOT event 1
timeout
timeout
GOT event 4
timeout
timeout
GOT event 10
.EE
.PP
Instead when the demo program runs using the emulation layer the output
is:
.IP
.EX
timeout
timeout
GOT event 1
timeout
timeout
GOT event 1
timeout
timeout
GOT event 4
timeout
timeout
GOT event 2011
.EE
.PP
In fact, the emulator uses a socketpair to generate the events.
Hangup is emulated by closing the other end of the socketpair: this
generates EPOLLHUP as well as EPOLLIN and EPOLLRDHUP.
.SH AUTHOR
VirtualSquare.
Project leader: Renzo Davoli.


vpoll.c
/*
 *   Virtual poll/select/epoll events
 *
 *   Copyright (C) 2019  Renzo Davoli <renzo@cs.unibo.it> VirtualSquare team.
 *
 *   This library is free software; you can redistribute it and/or modify it
 *   under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation; either version 2.1 of the License, or (at
 *   your option) any later version.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with this library; if not, write to the Free Software Foundation,
 *   Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdint.h>
#include <errno.h>
#include <sys/socket.h>
#include <sys/epoll.h>
#include <sys/eventfd.h>
#include <sys/ioctl.h>
#include <vpoll.h>
#include <fduserdata.h>

#define VPOLLDEV "/dev/vpoll"
#ifndef EFD_VPOLL
#define EFD_VPOLL (1 << 1)
#endif
#ifndef VPOLL_IOC_MAGIC
#define VPOLL_IOC_MAGIC '^'
#endif

static FDUSERDATA *fdtable = NULL;
static int vpolldev = 0;

/************************************ emulation mode ************************************/
static int vpollemu_create(uint32_t init_events, int flags);
static int vpollemu_close(int fd);
static int vpollemu_ctl(int fd, int op, uint32_t events);

/************************************ kernel EFD_VPOLL mode ************************************/

int vpoll_create(uint32_t init_events, int flags) {
	if (__builtin_expect(fdtable != NULL, 0))
		return vpollemu_create(init_events, flags);
	else if (vpolldev) {
		int fd = open(VPOLLDEV, O_RDWR | 
				(flags & FD_CLOEXEC ? O_CLOEXEC : 0));
		if (fd >= 0)
			ioctl(fd, _IO(VPOLL_IOC_MAGIC, VPOLL_CTL_ADDEVENTS), init_events);
		return fd;
	} else
		return eventfd(0, EFD_VPOLL |
				(flags & FD_CLOEXEC ? EFD_CLOEXEC : 0));
}

int vpoll_close(int fd) {
	if (__builtin_expect(fdtable != NULL, 0))
		return vpollemu_close(fd);
	else
	return close(fd);
}

int vpoll_ctl(int fd, int op, uint32_t events) {
	if (__builtin_expect(fdtable != NULL, 0))
		return vpollemu_ctl(fd, op, events);
	else if (vpolldev)
		return 0 - (ioctl(fd, _IO(VPOLL_IOC_MAGIC, op), events) < 0);
	else {
		uint64_t request = (((uint64_t) op) << 32) | events;
		return write(fd, &request, sizeof(request)) >= 0 ? 0 : -1;
	}
}

/************************************ emulation mode ************************************/
static inline void ignore_return(int v) {
}

static void emu_update_events(int fd, int datafd, uint32_t turnon, uint32_t turnoff) {
	char buf[4];
	//printf("%x %x\n", turnon, turnoff);
	if (turnon & EPOLLIN)
		ignore_return(write(datafd, "", 1));
	if (turnon & EPOLLOUT)
		ignore_return(read(datafd, buf, 4));
	if (turnoff & EPOLLIN)
		ignore_return(read(fd, buf, 4));
	if (turnoff & EPOLLOUT) {
		ignore_return(write(fd, "", 1));
		ignore_return(write(fd, "", 1));
	}
	if (turnon & EPOLLHUP)
		shutdown(datafd, SHUT_RDWR);
	else if (turnon & EPOLLRDHUP)
		shutdown(datafd, SHUT_WR);
}

static int vpollemu_create(uint32_t init_events, int flags) {
	int fds[2];
	int rv;
	int buflen=1;
	int *datafd;

	rv = socketpair(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK, 0, fds);
	if (rv < 0)
		return rv;
	if (flags & FD_CLOEXEC)
		fcntl(fds[0], F_SETFD, FD_CLOEXEC);
	fcntl(fds[1], F_SETFD, FD_CLOEXEC);
	setsockopt(fds[0], SOL_SOCKET, SO_SNDBUF, &buflen, sizeof(buflen));
	setsockopt(fds[1], SOL_SOCKET, SO_SNDBUF, &buflen, sizeof(buflen));
	datafd = fduserdata_new(fdtable, fds[0], int);
	if (datafd == NULL) {
		close(fds[0]);
		close(fds[1]);
		return -1;
	}
	*datafd = fds[1];
	emu_update_events(fds[0], fds[1], init_events & ~EPOLLOUT, ~init_events & EPOLLOUT);
	fduserdata_put(datafd);
	return fds[0];
}

static int vpollemu_close(int fd) {
	int *datafd = fduserdata_get(fdtable, fd);
	if (datafd == NULL)
		return -1;
	close(*datafd);
	close(fd);
	fduserdata_del(datafd);
	return 0;
}

static int vpollemu_ctl(int fd, int op, uint32_t events) {
	int *datafd = fduserdata_get(fdtable, fd);
	if (datafd == NULL)
		return -1;
	switch (op) {
		case VPOLL_CTL_ADDEVENTS: 
			emu_update_events(fd, *datafd, events, 0);
			break;
		case VPOLL_CTL_DELEVENTS: 
			emu_update_events(fd, *datafd, 0, events);
			break;
		case VPOLL_CTL_SETEVENTS: 
			emu_update_events(fd, *datafd, events, ~events);
			break;
		default: errno = EINVAL;
						 fduserdata_put(datafd);
						 return -1;
	}
	fduserdata_put(datafd);
	return 0;
}

/************************************ init/fini ************************************/
__attribute__((constructor))
	static void vpollemu_init() {
		/* use eventfd EFD_VPOLL when available */
		int testfd = eventfd(0, EFD_VPOLL | EFD_CLOEXEC);
		if (testfd >= 0)
			close(testfd);
		else {
		/* switch to device module mode */
			int testfd = open(VPOLLDEV, O_RDWR | O_CLOEXEC);
			if (testfd >= 0) {
				vpolldev = 1;
				close(testfd);
			} else
		/* switch to emulation mode if EFD_VPOLL or "/dev/vpoll" are not supported(yet). */
				fdtable = fduserdata_create(0);
		}
	}

__attribute__((destructor))
	static void vpollemu_fini() {
		if (fdtable != NULL)
			fduserdata_destroy(fdtable);
	}
